File: src/main/java/org/codeberg/zenxarch/zombies/entity/effect/util/StatusEffectInstanceBuilder.java
package org.codeberg.zenxarch.zombies.entity.effect.util;

import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.entity.effect.StatusEffect;
import net.minecraft.entity.effect.StatusEffectInstance;
import net.minecraft.registry.entry.RegistryEntry;
import net.minecraft.util.dynamic.Codecs;

public record StatusEffectInstanceBuilder(
    RegistryEntry<StatusEffect> effect, int duration, int amplifier) {
  public static final MapCodec<StatusEffectInstanceBuilder> CODEC =
      RecordCodecBuilder.mapCodec(
          instance ->
              instance
                  .group(
                      StatusEffect.ENTRY_CODEC
                          .fieldOf("effect")
                          .forGetter(StatusEffectInstanceBuilder::effect),
                      Codecs.POSITIVE_INT
                          .fieldOf("duration")
                          .forGetter(StatusEffectInstanceBuilder::duration),
                      Codecs.POSITIVE_INT
                          .fieldOf("amplifier")
                          .forGetter(StatusEffectInstanceBuilder::amplifier))
                  .apply(instance, StatusEffectInstanceBuilder::new));

  public static StatusEffectInstanceBuilder createInfinite(RegistryEntry<StatusEffect> effect) {
    return new StatusEffectInstanceBuilder(effect, -1, 1);
  }

  public static StatusEffectInstanceBuilder createInfinite(
      RegistryEntry<StatusEffect> effect, int amplifier) {
    return new StatusEffectInstanceBuilder(effect, -1, amplifier);
  }

  public static StatusEffectInstanceBuilder create(
      RegistryEntry<StatusEffect> effect, int duration) {
    return new StatusEffectInstanceBuilder(effect, duration, 1);
  }

  public static StatusEffectInstanceBuilder create(
      RegistryEntry<StatusEffect> effect, int duration, int amplifier) {
    return new StatusEffectInstanceBuilder(effect, duration, amplifier);
  }

  public StatusEffectInstance build() {
    return new StatusEffectInstance(effect, duration, amplifier);
  }
}

File: src/main/java/org/codeberg/zenxarch/zombies/entity/effect/util/StatusEffectInstanceBuilder.kt
package org.codeberg.zenxarch.zombies.entity.effect.util

import net.minecraft.entity.effect.StatusEffect
import net.minecraft.entity.effect.StatusEffectInstance
import net.minecraft.registry.entry.RegistryEntry

@JvmRecord
data class StatusEffectInstanceBuilder(val effect: RegistryEntry<StatusEffect>, val duration: Int, val amplifier: Int) {
    fun build(): StatusEffectInstance {
        return StatusEffectInstance(effect, duration, amplifier)
    }

    companion object {
        fun createInfinite(effect: RegistryEntry<StatusEffect>) =
            StatusEffectInstanceBuilder(effect, -1, 1)

        fun createInfinite(
            effect: RegistryEntry<StatusEffect>, amplifier: Int
        ) = StatusEffectInstanceBuilder(effect, -1, amplifier)

        fun create(
            effect: RegistryEntry<StatusEffect>, duration: Int
        ) = StatusEffectInstanceBuilder(effect, duration, 1)

        fun create(
            effect: RegistryEntry<StatusEffect>, duration: Int, amplifier: Int
        ) = StatusEffectInstanceBuilder(effect, duration, amplifier)
    }
}
